### PR TITLE
Testgen support for functions that return arrays.

### PR DESCRIPTION
**Pull request by NlightNFotis**
This PR introduces changes into cbmc designed to allow us to express code for C arrays under test, without which right now the behaviour we notice is completely bugged, producing code that is not syntactically valid C in the tests we generate.

### CODE REVIEW

File: src/ansi-c/expr2c.cpp [hunk 1]
```diff
@@ -860,6 +840,42 @@ std::string expr2ct::convert_array_type(
...
+    new_declarator+=declarator;
+
+    // Depending on precedences, we may add parentheses.
+    if(subtype_followed.id()==ID_code ||
+        (sizeof_nesting==0 &&
+         (subtype_followed.id()==ID_array ||
+          subtype_followed.id()==ID_incomplete_array)))
+      new_declarator="("+new_declarator+")";
+
+    if(subtype_followed.id()==ID_array)
```
Comment (thk123) [inline src/ansi-c/expr2c.cpp]: This behaviour should be explained and also toggable to ensure that the behaviour of `expr2ct` is not changed. In this case, I think a Boolean toggle passed in is sufficient.

File: src/ansi-c/expr2c.cpp [hunk 2]
```diff
@@ -860,6 +840,42 @@ std::string expr2ct::convert_array_type(
...
+std::string expr2ct::convert_pointer_type(
+  const typet &src,
+  const c_qualifierst &qualifiers,
+  const std::string &declarator)
+{
+    c_qualifierst new_qualifiers(qualifiers);
+    new_qualifiers.read(src);
+    std::string q=new_qualifiers.as_string();
+
+    c_qualifierst sub_qualifiers;
```
Comment (thk123) [inline src/ansi-c/expr2c.cpp]: Is this the right set of qualifiers if the modified behaviour is activated (for example, ensure we have tests for const pointers to const arrays, const pointers to non const arrays etc.)

File: src/ansi-c/expr2c.cpp [hunk 3]
```diff
@@ -860,6 +840,42 @@ std::string expr2ct::convert_array_type(
     src.subtype(), qualifiers, declarator_str+array_suffix);
 }
 
+std::string expr2ct::convert_pointer_type(
+  const typet &src,
+  const c_qualifierst &qualifiers,
+  const std::string &declarator)
+{
+    c_qualifierst new_qualifiers(qualifiers);
```
Comment (thk123) [inline src/ansi-c/expr2c.cpp]: This isn't in the code you are replacing, how come it is required here?

**Comment by thk123** [comment]
@NlightNFotis It'd be good to update the description since without the context of what this is for, this seems like a fairly random change!

File: src/ansi-c/expr2c.cpp (hunk 3 above)
Comment (NlightNFotis) [inline src/ansi-c/expr2c.cpp]: No, it is. If you see above, in the code block removed from if statement at line `491`, the code in the consequent statement is defining its own local `c_qualifierst` variable in order to call the method `read` on, as the relevant variable is still passed to the method that encloses that statement with a `const` qualifier, not allowing you to call `read` on it. If I don't do this, I get errors like 
```
/home/fotis/Devel/cbmc-public/src/ansi-c/expr2c.cpp:849: error: member function 'read' not viable: 'this' argument has type 'const c_qualifierst', but function is not marked const
    new_qualifiers.read(src);
    ^~~~~~~~~~~~~~
```

In other words it's required because we can't call the method on a const object that is referenced in the enclosing environment at both instances.

File: src/ansi-c/expr2c.cpp [hunk 4]
```diff
@@ -862,6 +842,58 @@ std::string expr2ct::convert_array_type(
...
+ Purpose: To generate a C-like type declaration of an pointer
+\*******************************************************************/
+
+std::string expr2ct::convert_pointer_type(
+  const typet &src,
+  const c_qualifierst &qualifiers,
+  const std::string &declarator,
+  bool toggle)
+{
+  c_qualifierst new_qualifiers(qualifiers);
```
Comment (thk123) [inline src/ansi-c/expr2c.cpp]: I am still confused about this - I thought we worked out it wasn't needed. I'm not really sure what the difference between `qualifiers` and `new_qualifiers`. The old `q` seems to be generated from the `qualifiers` that is passed in.

**Comment by NlightNFotis** [comment]
@tautschnig Yes, I am sorry the commit message was not clear enough. I had published this PR early in an attempt to get reviews faster, and the ambiguous commit message kinda got stuck with it until now.

The essence of the change is to allow us to have better support for generating assertions (well, at least generate syntactically valid C assertions) for members of heap allocated array-like structures. An example of this would be something like this:

```
int *fun()
{
  int *a = malloc(sizeof(int)*10);
  for (int i = 0; i < 10; i++) {
    a[i] = i;
  }
  
  a[5] = a[6] * 2;
  return a;
}
```

With this patch, in combination with another one that accompanies it in test-gen, we can generate syntactically valid C assertions for all members of this heap allocated entry, in an array index accessing fashion, to be more recognisable to C programmers, whereas without it, we generate code that can't compile.

**Comment by tautschnig** [comment]
It would help a lot to have a regression test included in this pull request. I can understand your words, but it's very hard to match the code changes to this.

**Comment by NlightNFotis** [comment]
There are regression tests, but they are on the internal test-gen repository. After discussion with one of my colleagues, my understanding is that there is no public executable using `expr2c`. How do you propose I write unit tests for this?

**Comment by tautschnig** [comment]
`goto-instrument --dump-c` heavily uses `expr2ct`, but even any `--show-goto-functions` will use `expr2ct`. If you add them in the `goto-instrument` regression test directory then `chain.sh` will nicely exercise all that for you.

**Comment by NlightNFotis** [comment]
@tautschnig After a discussion with my colleague, we concluded that the behaviour we wanted to fix is something that isn't manifested in cbmc, but it manifests in test-gen. So we decided to go ahead with a test that demonstrates we have broken nothing in our change to cbmc for better test-gen support. Please tell me if that's close to what you would expect.

File: regression/goto-instrument/array_01/test.desc [hunk 5]
```diff
@@ -0,0 +1,18 @@
+CORE
+main.c
+--dump-c 
+^        signed int \*\*Array;$
+^        void \*return_value_malloc\$1;$
+^        return_value_malloc\$1=malloc\(\(signed int\)\(2ul \* sizeof\(signed int \*\)
```
Comment (tautschnig) [inline regression/goto-instrument/array_01/test.desc]: Those type casts to `signed int` aren't portable - you'll need `long` types here (as well) as the failing regression tests indicate.

**Comment by tautschnig** [comment]
> With this patch, in combination with another one that accompanies it in test-gen, we can generate syntactically valid C assertions for all members of this heap allocated entry, in an array index accessing fashion, to be more recognisable to C programmers, whereas without it, we generate code that can't compile.

Would it be possible to share some of this output here? Also, I don't see why this is even related to test-gen - generating assertions does not seem to be a matter of test-(input) generation!?

File: src/ansi-c/expr2c.cpp [hunk 6]
```diff
@@ -862,6 +842,55 @@ std::string expr2ct::convert_array_type(
...
+  new_declarator+=declarator;
+
+  // Depending on precedences, we may add parentheses.
+  if(subtype_followed.id()==ID_code ||
+      (sizeof_nesting==0 &&
+        (subtype_followed.id()==ID_array ||
+          subtype_followed.id()==ID_incomplete_array)))
+      new_declarator="("+new_declarator+")";
+
+  if(subtype_followed.id()==ID_array && toggle)
```
Comment (peterschrammel) [inline src/ansi-c/expr2c.cpp:885]: Separate the refactoring from your actual change.

File: src/ansi-c/expr2c.cpp [hunk 7]
```diff
@@ -862,6 +842,55 @@ std::string expr2ct::convert_array_type(
...
+  Inputs:
+          src - The pointer type to convert
+          qualifiers - A list of the qualifiers of the pointer declaration
+          declarator_str - The declaration string we will build
+          toggle - Activate the pointer -> array treating behavior
+
+ Outputs: A C-like type declaration of a pointer
+
+ Purpose: To generate a C-like type declaration of an pointer
+\*******************************************************************/
```
Comment (peterschrammel) [inline src/ansi-c/expr2c.cpp:856]: doxygen

File: src/ansi-c/expr2c.cpp [hunk 8]
```diff
@@ -862,6 +842,55 @@ std::string expr2ct::convert_array_type(
 
 /*******************************************************************\
 
+Function: expr2ct::convert_pointer_type
```
Comment (peterschrammel) [inline src/ansi-c/expr2c.cpp:845]: The commit message does not describe the change.

File: src/ansi-c/expr2c.cpp [hunk 9]
```diff
@@ -862,6 +842,55 @@ std::string expr2ct::convert_array_type(
...
+  // Depending on precedences, we may add parentheses.
+  if(subtype_followed.id()==ID_code ||
+      (sizeof_nesting==0 &&
+        (subtype_followed.id()==ID_array ||
+          subtype_followed.id()==ID_incomplete_array)))
+      new_declarator="("+new_declarator+")";
+
+  if(subtype_followed.id()==ID_array && toggle)
+    return convert_rec(src.subtype().subtype(), sub_qualifiers,
+      new_declarator.substr(1, new_declarator.size()-2));
```
Comment (peterschrammel) [inline src/ansi-c/expr2c.cpp:887]: This doesn't seem the right way of doing that.
